src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const unsigned num) noexcept
+{
+    char tmp[UINT_MAX_CHARS + 1];
+    int len = snprintf(tmp, UINT_MAX_CHARS + 1, "%d", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const unsigned num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const long num) noexcept
+{
+    char tmp[LONG_MAX_CHARS + 1];
+    int len = snprintf(tmp, LONG_MAX_CHARS + 1, "%ld", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const long num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const unsigned long num) noexcept
+{
+    char tmp[ULONG_MAX_CHARS + 1];
+    int len = snprintf(tmp, ULONG_MAX_CHARS + 1, "%ld", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const unsigned long num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const endline &) noexcept
+{
+    int r = put("\n", 1);
+    r += flush_nx();
+    return (r >= 1);
+}
+
+bool ostream::write(const endline &)
+{
+    bool r = write_nx(endl);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const flushbuf &) noexcept
+{
+    return (flush_nx());
+}
+
+bool ostream::write(const flushbuf &)
+{
+    return (flush());
+}
+
+ssize_t ostream::write_buf_nx(const char *msg) noexcept
+{
+    return put(msg, strlen(msg));
+}
+
+ssize_t ostream::write_buf(const char *msg)
+{
+    ssize_t r = write_buf_nx(msg);
+    if (r < 0) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+ssize_t ostream::write_buf_nx(const std::string &msg) noexcept
+{
+    return put(msg.c_str(), msg.size());
+}
+
+ssize_t ostream::write_buf(const std::string &msg)
+{
+    ssize_t r = write_buf_nx(msg);
+    if (r < 0) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+ostream::operator bool() noexcept
+{
+    return good();
+}
+
+ostream::~ostream() noexcept
+{
+    // We don't use close()/close_nx() functions because we cannot do anything about flush_nx() failure
+    // in this stage.
```

Comment:
If you fix `close_nx` so that it always closes the file descriptor, you can use it.